### PR TITLE
DO NOT MERGE: Lock heartbeat - discuss

### PR DESCRIPTION
**Pull request by mxmCherry**

### CODE REVIEW

File: lib/fleiss/backend/active_record/concern.rb [hunk 1]
```diff
@@ -114,6 +127,40 @@ def with_isolation(&block)
             yield
           end
         end
+
+        def start_heartbeat(owner)
+          raise "Multiple start_heartbeat for job id=#{id}" if @heartbeat # indicates that something is very wrong with worker
+
+          interval = try(:heartbeat_interval) || return
+
+          @heartbeat = Concurrent::TimerTask.new(execution_interval: interval) do |_task|
```
Comment (mxmCherry) [inline lib/fleiss/backend/active_record/concern.rb:136]: This is the heavy bit - creating a timer task for each job seems quite wasteful, right?

Comment (mxmCherry) [reply lib/fleiss/backend/active_record/concern.rb:136]: Ideally, this should be done not even per worker, but per worker thread, somewhere in Fleiss::Executor.

We can "remember" `owner` in Fleiss::Worker and pass it to Fleiss::Executor.

Then, in Executor, we can schedule a Concurrent::TimerTask to refresh locks like `UPDATE lock_expires_at = CURRENT_TIMESTAMP + $ttl WHERE owner = $owner`.

Comment (mxmCherry) [reply lib/fleiss/backend/active_record/concern.rb:136]: More of a :duck: PR, but can you review (comments as well) @dim ?

Comment (mxmCherry) [reply lib/fleiss/backend/active_record/concern.rb:136]: Btw, it's not easy to just plug heartbeat into Executor as well.

Executor is based on: https://github.com/ruby-concurrency/concurrent-ruby/blob/master/lib/concurrent-ruby/concurrent/executor/simple_executor_service.rb

And authors warn of it's quite poor performance: https://ruby-concurrency.github.io/concurrent-ruby/master/Concurrent/SimpleExecutorService.html

> This is perhaps the most inefficient executor service in this library. It exists mainly for testing an debugging. Thread creation and management is expensive in Ruby and this executor performs no resource pooling. This can be very beneficial during testing and debugging because it decouples the using code from the underlying executor implementation. In production this executor will likely lead to suboptimal performance.

The main problem is this: https://github.com/ruby-concurrency/concurrent-ruby/blob/master/lib/concurrent-ruby/concurrent/executor/simple_executor_service.rb#L36-L50

It's basically "create thread on .post".

So basically, we cannot just identify owner once and start a heartbeat task to refresh locks.

With this approach "create thread on .post" it makes sense to keep what I've done - basically, per-job heartbeat.

Otherwise it'll drag too many changes to fleiss - basically, swapping workers to smth more performant (maybe proper thread pool or so), but even then - we'll have to dig into how to "hook" into thread lifecycle (started/stopped), which I'm not even sure is easy/implemented.

Comment (mxmCherry) [reply lib/fleiss/backend/active_record/concern.rb:136]: Maybe the best option would be to:

- look into something better than SimpleExecutorService, smth like thread pool
- ideally, have own (subclassed?) Thread with all the hooks implemented to plug in heartbeat-ing
